sponsorship/src/main/java/com/kyriba/conference/sponsorship/api/PlanController.java

```diff
@@ -29,55 +38,69 @@
  */
 @RestController
 @RequiredArgsConstructor
-@RequestMapping(value = "${api.version}/sponsorship/plans",
+@RequestMapping(value = "${api.version.path}/plans",
     consumes = { MediaType.APPLICATION_JSON_UTF8_VALUE }, produces = { MediaType.APPLICATION_JSON_UTF8_VALUE })
 @Api(value = "Register a new sponsorship plan")
 public class PlanController
 {
   private final PlanService planService;
 
 
-  @SuppressWarnings("unused")
   @ApiOperation(value = "Register the plan")
-  @PostMapping
   @ApiResponses(value = {
-      @ApiResponse(code = 200, message = "OK", response = PlanRegistered.class),
-      @ApiResponse(code = 401, message = "Failed to register the plan", response = PlanRegistered.class)
+      @ApiResponse(code = 200, message = "OK"),
+      @ApiResponse(code = 401, message = "Failed to register the plan")
   })
-  PlanRegistered register(@Valid @RequestBody PlanRegistrationRequest request)
+  @PostMapping
+  PlanRegistrationResponse register(@Valid @RequestBody PlanRegistrationRequest request)
   {
-    final String randomId = "234";
-    final Plan plan = Plan.builder()
-        .id(randomId)
-        .category(request.getCategory())
-        .sponsorId(request.getSponsorEmail())
-        .build();
-    return new PlanRegistered(plan.getId());
+    return new PlanRegistrationResponse(planService.createPlan(request.getCategory(), request.getSponsorEmail()));
   }
 
 
-  @SuppressWarnings("unused")
   @ApiOperation(value = "Cancel the plan")
-  @PutMapping("/{id}/cancellation")
   @ApiResponses(value = {
-      @ApiResponse(code = 200, message = "OK", response = PlanCancellationResponse.class),
-      @ApiResponse(code = 401, message = "Failed to cancel the plan", response = PlanCancellationResponse.class)
+      @ApiResponse(code = 200, message = "OK"),
+      @ApiResponse(code = 401, message = "Failed to cancel the plan")
   })
-  PlanCancellationResponse cancel(@ApiParam(value = "Id of the plan to cancel", required = true) @PathVariable String id)
+  @DeleteMapping("/{id}")
+  void cancel(@ApiParam(value = "Id of the plan to cancel", required = true) @PathVariable Long id)
   {
-    return new PlanCancellationResponse(id);
+    planService.deletePlan(id);

```

Comment:
how do you process case when the plan is used by sponsor and cannot be removed?

Reply:
A plan has a reference to a sponsor, not vise versa, so it is not a problem

Reply:
Strange reference. Logically sponsor participate in event using one of proposed plans, so relation owning entity here is sponsor.

Reply:
Proposed plans represented by PlanCategory enum. A plan of a selected category and associated with a specified sponsor is represented by Plan entity.  There is no dedicated entity or table for proposed plans, those are plans where sponsorID == null